In [172]:
import pandas as pd
import numpy as np
import matplotlib as plt
%matplotlib inline
import seaborn as sns
import sklearn

url = 'D:\\zoom\Python\\input\\adult.csv'

Data Pre Processing

In [173]:
df = pd.read_csv(url,na_values=['#NAME?'])

In [174]:
df.head()

age         workclass    fnlwgt  education  education_num  \
0  39.0         State-gov   77516.0  Bachelors           13.0   
1  50.0  Self-emp-not-inc   83311.0  Bachelors           13.0   
2  38.0           Private  215646.0    HS-grad            9.0   
3  53.0           Private  234721.0       11th            7.0   
4  28.0           Private  338409.0  Bachelors           13.0   

       marital_status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black     NaN   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital_gain  capital_loss  hours_per_week native_country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [175]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             4952 non-null   float64
 1   workclass       5000 non-null   object 
 2   fnlwgt          4893 non-null   float64
 3   education       5000 non-null   object 
 4   education_num   4943 non-null   float64
 5   marital_status  5000 non-null   object 
 6   occupation      5000 non-null   object 
 7   relationship    5000 non-null   object 
 8   race            4736 non-null   object 
 9   sex             4953 non-null   object 
 10  capital_gain    5000 non-null   int64  
 11  capital_loss    5000 non-null   int64  
 12  hours_per_week  5000 non-null   int64  
 13  native_country  5000 non-null   object 
 14  income          5000 non-null   object 
dtypes: float64(3), int64(3), object(9)
memory usage: 586.1+ KB


Age

In [176]:
df['age'].unique()

array([39., 50., 38., 53., 28., 37., 49., 52., 31., 42., 30., 23., 32.,
       40., 34., nan, 43., 54., 35., 59., 56., 19., 20., 45., 22., 48.,
       21., 24., 25., 57., 44., 41., 29., 18., 47., 46., 36., 79., 27.,
       67., 33., 76., 17., 55., 61., 70., 64., 71., 68., 66., 51., 58.,
       26., 60., 90., 75., 65., 77., 62., 63., 80., 72., 74., 69., 73.,
       81., 78., 88., 82.])

In [177]:
df['age'].isna().sum()

48

In [178]:
df['age'].describe()

count    4952.000000
mean       38.586430
std        13.582256
min        17.000000
25%        28.000000
50%        37.000000
75%        47.000000
max        90.000000
Name: age, dtype: float64

In [179]:
df['age'].fillna(df['age'].mean(),inplace=True)

In [180]:
df['age'] = df['age'].astype(int)

Work Class

In [181]:
df['workclass'].value_counts()

Private             3435
Self-emp-not-inc     383
?                    331
Local-gov            329
State-gov            193
Self-emp-inc         182
Federal-gov          146
Without-pay            1
Name: workclass, dtype: int64

In [182]:
df.loc[df['workclass']=='?','workclass']='Student'

Converting all '?' to student work class because when checked on occupation there also same number of '?' and when checked on education majority are from some collage, HS-grad and Bachelors 

In [183]:
df['workclass'] = df['workclass'].astype('category')

Fnlegt

In [184]:
df['fnlwgt'].isna().sum()

107

In [185]:
df = df.dropna(subset=['fnlwgt'])

Education

In [186]:
df['education'].isna().sum()

0

In [187]:
df['education'].unique()

array(['Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Some-college',
       'Assoc-acdm', 'Assoc-voc', '7th-8th', 'Doctorate', '?',
       'Prof-school', '5th-6th', '10th', '1st-4th', 'Preschool', '12th'],
      dtype=object)

In [188]:
df['education'].value_counts()

HS-grad         1560
Some-college    1088
Bachelors        810
Masters          242
Assoc-voc        210
11th             196
Assoc-acdm       160
10th             141
7th-8th           94
Prof-school       84
9th               73
?                 57
Doctorate         54
12th              53
5th-6th           42
1st-4th           22
Preschool          7
Name: education, dtype: int64

In [189]:
df.loc[df['education']=='?','education']='unknown'

In [190]:
df['education'] = df['education'].astype('category')

Education_num

In [191]:
df['education_num'].isna().sum()

57

In [192]:
df.dropna(subset=['education_num'],inplace=True)

In [193]:
df['education_num'].value_counts()

9.0     1560
10.0    1088
13.0     810
14.0     242
11.0     210
7.0      196
12.0     160
6.0      141
4.0       94
15.0      84
5.0       73
16.0      54
8.0       53
3.0       42
2.0       22
1.0        7
Name: education_num, dtype: int64

Marital_status

In [194]:
df['marital_status'].isna().sum()

0

In [195]:
df['marital_status'].value_counts()

Married-civ-spouse       2180
Never-married            1619
Divorced                  658
Separated                 156
Widowed                   155
Married-spouse-absent      63
Married-AF-spouse           5
Name: marital_status, dtype: int64

Occupation

In [196]:
df['occupation'].isna().sum()

0

In [197]:
df['occupation'].value_counts()

Prof-specialty       600
Exec-managerial      599
Craft-repair         595
Sales                567
Adm-clerical         560
Other-service        483
?                    321
Machine-op-inspct    300
Transport-moving     243
Handlers-cleaners    192
Farming-fishing      137
Tech-support         135
Protective-serv       84
Priv-house-serv       18
Armed-Forces           2
Name: occupation, dtype: int64

In [198]:
df.loc[df['occupation'] =='?','occupation']='Student'

In [199]:
df['occupation'] = df['occupation'].astype('category')

Relationship

In [200]:
df['relationship'].isna().sum()

0

In [201]:
df['relationship'].value_counts()

Husband           1919
Not-in-family     1264
Own-child          749
Unmarried          523
Wife               237
Other-relative     144
Name: relationship, dtype: int64

Race

In [202]:
df['race'].isna().sum()

251

In [203]:
df['race'].unique()

array(['White', 'Black', 'Asian-Pac-Islander', 'Amer-Indian-Eskimo', nan,
       'Other'], dtype=object)

In [204]:
df['race'].value_counts()

White                 3894
Black                  475
Asian-Pac-Islander     141
Amer-Indian-Eskimo      46
Other                   29
Name: race, dtype: int64

In [205]:
df['race'].fillna('Other',inplace=True)

Sex

In [206]:
df['sex'].isna().sum()

47

In [207]:
df['sex'].unique()

array(['Male', nan, 'Female'], dtype=object)

In [208]:
df.dropna(subset=['sex'],inplace=True)

Capital_gain

In [209]:
df['capital_gain'].isna().sum()

0

In [210]:
df['capital_gain'].describe()

count     4789.000000
mean      1060.760702
std       7194.824907
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max      99999.000000
Name: capital_gain, dtype: float64

Capital_loss

In [211]:
df['capital_loss'].isna().sum()

0

In [212]:
df['capital_loss'].describe()

count    4789.000000
mean       92.214241
std       407.782383
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max      2547.000000
Name: capital_loss, dtype: float64

Hours_per_week

In [213]:
df['hours_per_week'].isna().sum()

0

In [214]:
df['hours_per_week'].unique()

array([40, 13, 16, 45, 50, 80, 30, 60, 20, 52, 44, 15, 25, 38, 43, 35, 55,
       48, 58, 32, 70,  2, 22, 56, 41, 28, 36, 24, 46, 42, 12, 65,  1, 10,
       34, 75, 98, 33, 54,  8,  6, 64, 19, 18, 72,  5,  9, 47, 37, 21, 26,
       14,  4, 59,  7, 99, 53, 39, 62, 57, 78, 90, 66, 11, 49, 84,  3, 17,
       68, 27, 85, 31, 51, 77], dtype=int64)

Native_country

In [215]:
df['native_country'].unique()

array(['United-States', 'Cuba', 'Jamaica', 'India', '?', 'Mexico',
       'South', 'Puerto-Rico', 'Honduras', 'England', 'Canada', 'Germany',
       'Iran', 'Philippines', 'Italy', 'Poland', 'Columbia', 'Cambodia',
       'Thailand', 'Laos', 'Taiwan', 'Haiti', 'Portugal',
       'Dominican-Republic', 'El-Salvador', 'France', 'Guatemala',
       'China', 'Japan', 'Yugoslavia', 'Peru',
       'Outlying-US(Guam-USVI-etc)', 'Scotland', 'Trinadad&Tobago',
       'Greece', 'Nicaragua', 'Vietnam', 'Hong', 'Ireland', 'Ecuador'],
      dtype=object)

In [216]:
df.loc[df['native_country']=='?','native_country']='Others'

Income

In [217]:
df['income'].unique()

array(['<=50K', '>50K'], dtype=object)

In [218]:
df = df.astype({'marital_status':'category','occupation':'category','relationship':'category','race':'category',
                'sex':'category','native_country':'category'})

In [219]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4789 entries, 0 to 4999
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   age             4789 non-null   int32   
 1   workclass       4789 non-null   category
 2   fnlwgt          4789 non-null   float64 
 3   education       4789 non-null   category
 4   education_num   4789 non-null   float64 
 5   marital_status  4789 non-null   category
 6   occupation      4789 non-null   category
 7   relationship    4789 non-null   category
 8   race            4789 non-null   category
 9   sex             4789 non-null   category
 10  capital_gain    4789 non-null   int64   
 11  capital_loss    4789 non-null   int64   
 12  hours_per_week  4789 non-null   int64   
 13  native_country  4789 non-null   category
 14  income          4789 non-null   object  
dtypes: category(8), float64(2), int32(1), int64(3), object(1)
memory usage: 322.3+ KB


Data Transformation

    Encoding

In [220]:
census_data = pd.get_dummies(df[['workclass','education','marital_status','occupation',
                                      'relationship','race','sex','native_country']])

In [221]:
data = pd.concat((df.drop(['workclass','education','marital_status','occupation',
                                'relationship','race','sex','native_country'], axis = 1),census_data),axis = 1)

In [223]:
y= data.pop('income')

In [224]:
data.head()

age    fnlwgt  education_num  capital_gain  capital_loss  hours_per_week  \
0   39   77516.0           13.0          2174             0              40   
1   50   83311.0           13.0             0             0              13   
2   38  215646.0            9.0             0             0              40   
4   28  338409.0           13.0             0             0              40   
5   37  284582.0           14.0             0             0              40   

   workclass_Federal-gov  workclass_Local-gov  workclass_Private  \
0                      0                    0                  0   
1                      0                    0                  0   
2                      0                    0                  1   
4                      0                    0                  1   
5                      0                    0                  1   

   workclass_Self-emp-inc  ...  native_country_Portugal  \
0                       0  ...                        0   
1                       0  ...                        0   
2                       0  ...                        0   
4                       0  ...                        0   
5                       0  ...                        0   

   native_country_Puerto-Rico  native_country_Scotland  native_country_South  \
0                           0                        0                     0   
1                           0                        0                     0   
2                           0                        0                     0   
4                           0                        0                     0   
5                           0                        0                     0   

   native_country_Taiwan  native_country_Thailand  \
0                      0                        0   
1                      0                        0   
2                      0                        0   
4                      0                        0   
5                      0                        0   

   native_country_Trinadad&Tobago  native_country_United-States  \
0                               0                             1   
1                               0                             1   
2                               0                             1   
4                               0                             0   
5                               0                             1   

   native_country_Vietnam  native_country_Yugoslavia  
0                       0                          0  
1                       0                          0  
2                       0                          0  
4                       0                          0  
5                       0                          0  

[5 rows x 106 columns]

    Standardazing Data

In [226]:
from sklearn.preprocessing import StandardScaler

In [227]:
scale = StandardScaler()

In [230]:
data_std = scale.fit_transform(data)

Modelling

    Data Splitting 

In [232]:
from sklearn.model_selection import train_test_split

In [236]:
X_train, X_test, y_train, y_test = train_test_split(data_std, y, test_size=0.2)

Logistic Model

In [237]:
from sklearn.linear_model import LogisticRegression

In [238]:
logic = LogisticRegression()

In [239]:
logic.fit(X_train,y_train)

LogisticRegression()

In [241]:
logic.coef_

array([[ 0.38368962,  0.0529009 ,  0.31099362,  2.05425033,  0.27835818,
         0.42932879,  0.14824797, -0.0434411 ,  0.03076641,  0.06874008,
        -0.14041125,  0.02275616, -0.03357952,  0.0031417 , -0.12442388,
        -0.03808616, -0.03069807, -0.03878004, -0.01144269, -0.06835577,
        -0.1992201 ,  0.03286582,  0.04891627,  0.14406364,  0.18621816,
        -0.00443243,  0.064857  , -0.12487756,  0.10196429, -0.09827376,
         0.        , -0.19985077, -0.14353797,  0.86448746, -0.07228922,
        -0.63715804, -0.13224618, -0.13931459, -0.10123994, -0.08606455,
         0.04113255,  0.32318211, -0.26655075, -0.14764154, -0.08164074,
        -0.27458239, -0.19454583,  0.21430157,  0.02583945,  0.07127532,
        -0.03357952,  0.11824935, -0.0308427 , -0.06775046,  0.25070923,
         0.02289189, -0.48532044,  0.08713558,  0.31228452, -0.12006531,
         0.11160676,  0.02111267, -0.10091668,  0.02562077, -0.16620563,
         0.16620563,  0.06770448,  0.07323815, -0.0

In [242]:
logic.intercept_

array([-2.25544045])

In [245]:
predict_logic =logic.predict(X_test)

    Accuracy test for Logistic model

In [247]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [248]:
print(accuracy_score(y_test,predict_logic))

0.8298538622129437


In [250]:
print(confusion_matrix(y_test,predict_logic))

[[670  73]
 [ 90 125]]


Decision Tree

In [251]:
from sklearn.tree import DecisionTreeClassifier

In [252]:
decision  = DecisionTreeClassifier()

In [253]:
decision.fit(X_train,y_train)

DecisionTreeClassifier()

In [255]:
predict_decision = decision.predict(X_test)

    Accuracy test for Decision Tree

In [256]:
print(accuracy_score(y_test,predict_decision))

0.7943632567849687


In [257]:
print(confusion_matrix(y_test,predict_decision))

[[630 113]
 [ 84 131]]


Random Forest

In [259]:
from sklearn.ensemble import RandomForestClassifier

In [260]:
rf = RandomForestClassifier()

In [261]:
rf.fit(X_train,y_train)

RandomForestClassifier()

In [262]:
predict_rf = rf.predict(X_test)

     Accuracy test for Random Forest

In [263]:
print(accuracy_score(y_test,predict_rf))

0.8308977035490606


In [264]:
print(confusion_matrix(y_test,predict_rf))

[[669  74]
 [ 88 127]]


In [265]:
print(classification_report(y_test,predict_rf))

              precision    recall  f1-score   support

       <=50K       0.88      0.90      0.89       743
        >50K       0.63      0.59      0.61       215

    accuracy                           0.83       958
   macro avg       0.76      0.75      0.75       958
weighted avg       0.83      0.83      0.83       958



Grid Search for Random Forest

In [279]:
from sklearn.model_selection import GridSearchCV

In [280]:
params ={'n_estimators':[150,200,250,300],'max_depth':[5,10,15,20,25,30],'min_samples_leaf':[2,3,4,5]}

In [282]:
grid = GridSearchCV(rf,params,verbose=3)

In [283]:
grid.fit(X_train,y_train)

Fitting 5 folds for each of 96 candidates, totalling 480 fits
[CV] max_depth=5, min_samples_leaf=2, n_estimators=150 ...............


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  max_depth=5, min_samples_leaf=2, n_estimators=150, score=0.849, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=150 ...............


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.2s remaining:    0.0s


[CV]  max_depth=5, min_samples_leaf=2, n_estimators=150, score=0.833, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=150 ...............


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.4s remaining:    0.0s


[CV]  max_depth=5, min_samples_leaf=2, n_estimators=150, score=0.846, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=150 ...............
[CV]  max_depth=5, min_samples_leaf=2, n_estimators=150, score=0.836, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=150 ...............
[CV]  max_depth=5, min_samples_leaf=2, n_estimators=150, score=0.834, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=200 ...............
[CV]  max_depth=5, min_samples_leaf=2, n_estimators=200, score=0.854, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=200 ...............
[CV]  max_depth=5, min_samples_leaf=2, n_estimators=200, score=0.829, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=200 ...............
[CV]  max_depth=5, min_samples_leaf=2, n_estimators=200, score=0.845, total=   0.3s
[CV] max_depth=5, min_samples_leaf=2, n_estimators=200 ...............
[CV]  max_depth=5, min_samples_leaf=2, n_estimators=200, score=0.833, 

[CV]  max_depth=5, min_samples_leaf=4, n_estimators=300, score=0.851, total=   0.5s
[CV] max_depth=5, min_samples_leaf=4, n_estimators=300 ...............
[CV]  max_depth=5, min_samples_leaf=4, n_estimators=300, score=0.832, total=   0.5s
[CV] max_depth=5, min_samples_leaf=4, n_estimators=300 ...............
[CV]  max_depth=5, min_samples_leaf=4, n_estimators=300, score=0.851, total=   0.5s
[CV] max_depth=5, min_samples_leaf=4, n_estimators=300 ...............
[CV]  max_depth=5, min_samples_leaf=4, n_estimators=300, score=0.839, total=   0.5s
[CV] max_depth=5, min_samples_leaf=4, n_estimators=300 ...............
[CV]  max_depth=5, min_samples_leaf=4, n_estimators=300, score=0.830, total=   0.5s
[CV] max_depth=5, min_samples_leaf=5, n_estimators=150 ...............
[CV]  max_depth=5, min_samples_leaf=5, n_estimators=150, score=0.847, total=   0.2s
[CV] max_depth=5, min_samples_leaf=5, n_estimators=150 ...............
[CV]  max_depth=5, min_samples_leaf=5, n_estimators=150, score=0.829, 

[CV]  max_depth=10, min_samples_leaf=3, n_estimators=200, score=0.858, total=   0.4s
[CV] max_depth=10, min_samples_leaf=3, n_estimators=200 ..............
[CV]  max_depth=10, min_samples_leaf=3, n_estimators=200, score=0.850, total=   0.4s
[CV] max_depth=10, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=10, min_samples_leaf=3, n_estimators=250, score=0.855, total=   0.5s
[CV] max_depth=10, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=10, min_samples_leaf=3, n_estimators=250, score=0.847, total=   0.5s
[CV] max_depth=10, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=10, min_samples_leaf=3, n_estimators=250, score=0.859, total=   0.5s
[CV] max_depth=10, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=10, min_samples_leaf=3, n_estimators=250, score=0.855, total=   0.5s
[CV] max_depth=10, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=10, min_samples_leaf=3, n_estimators=250, score=

[CV]  max_depth=15, min_samples_leaf=2, n_estimators=150, score=0.855, total=   0.4s
[CV] max_depth=15, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=15, min_samples_leaf=2, n_estimators=150, score=0.856, total=   0.4s
[CV] max_depth=15, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=15, min_samples_leaf=2, n_estimators=150, score=0.855, total=   0.4s
[CV] max_depth=15, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=15, min_samples_leaf=2, n_estimators=150, score=0.855, total=   0.3s
[CV] max_depth=15, min_samples_leaf=2, n_estimators=200 ..............
[CV]  max_depth=15, min_samples_leaf=2, n_estimators=200, score=0.864, total=   0.5s
[CV] max_depth=15, min_samples_leaf=2, n_estimators=200 ..............
[CV]  max_depth=15, min_samples_leaf=2, n_estimators=200, score=0.850, total=   0.5s
[CV] max_depth=15, min_samples_leaf=2, n_estimators=200 ..............
[CV]  max_depth=15, min_samples_leaf=2, n_estimators=200, score=

[CV]  max_depth=15, min_samples_leaf=4, n_estimators=250, score=0.851, total=   0.6s
[CV] max_depth=15, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=15, min_samples_leaf=4, n_estimators=300, score=0.855, total=   0.7s
[CV] max_depth=15, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=15, min_samples_leaf=4, n_estimators=300, score=0.847, total=   0.7s
[CV] max_depth=15, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=15, min_samples_leaf=4, n_estimators=300, score=0.858, total=   0.7s
[CV] max_depth=15, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=15, min_samples_leaf=4, n_estimators=300, score=0.852, total=   0.7s
[CV] max_depth=15, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=15, min_samples_leaf=4, n_estimators=300, score=0.847, total=   0.7s
[CV] max_depth=15, min_samples_leaf=5, n_estimators=150 ..............
[CV]  max_depth=15, min_samples_leaf=5, n_estimators=150, score=

[CV]  max_depth=20, min_samples_leaf=3, n_estimators=200, score=0.856, total=   0.5s
[CV] max_depth=20, min_samples_leaf=3, n_estimators=200 ..............
[CV]  max_depth=20, min_samples_leaf=3, n_estimators=200, score=0.855, total=   0.5s
[CV] max_depth=20, min_samples_leaf=3, n_estimators=200 ..............
[CV]  max_depth=20, min_samples_leaf=3, n_estimators=200, score=0.850, total=   0.5s
[CV] max_depth=20, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=20, min_samples_leaf=3, n_estimators=250, score=0.867, total=   0.6s
[CV] max_depth=20, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=20, min_samples_leaf=3, n_estimators=250, score=0.851, total=   0.6s
[CV] max_depth=20, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=20, min_samples_leaf=3, n_estimators=250, score=0.866, total=   0.6s
[CV] max_depth=20, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=20, min_samples_leaf=3, n_estimators=250, score=

[CV]  max_depth=25, min_samples_leaf=2, n_estimators=150, score=0.867, total=   0.4s
[CV] max_depth=25, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=25, min_samples_leaf=2, n_estimators=150, score=0.846, total=   0.4s
[CV] max_depth=25, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=25, min_samples_leaf=2, n_estimators=150, score=0.855, total=   0.4s
[CV] max_depth=25, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=25, min_samples_leaf=2, n_estimators=150, score=0.856, total=   0.4s
[CV] max_depth=25, min_samples_leaf=2, n_estimators=150 ..............
[CV]  max_depth=25, min_samples_leaf=2, n_estimators=150, score=0.850, total=   0.4s
[CV] max_depth=25, min_samples_leaf=2, n_estimators=200 ..............
[CV]  max_depth=25, min_samples_leaf=2, n_estimators=200, score=0.862, total=   0.6s
[CV] max_depth=25, min_samples_leaf=2, n_estimators=200 ..............
[CV]  max_depth=25, min_samples_leaf=2, n_estimators=200, score=

[CV]  max_depth=25, min_samples_leaf=4, n_estimators=250, score=0.860, total=   0.6s
[CV] max_depth=25, min_samples_leaf=4, n_estimators=250 ..............
[CV]  max_depth=25, min_samples_leaf=4, n_estimators=250, score=0.850, total=   0.6s
[CV] max_depth=25, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=25, min_samples_leaf=4, n_estimators=300, score=0.862, total=   0.7s
[CV] max_depth=25, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=25, min_samples_leaf=4, n_estimators=300, score=0.851, total=   0.7s
[CV] max_depth=25, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=25, min_samples_leaf=4, n_estimators=300, score=0.855, total=   0.7s
[CV] max_depth=25, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=25, min_samples_leaf=4, n_estimators=300, score=0.858, total=   0.7s
[CV] max_depth=25, min_samples_leaf=4, n_estimators=300 ..............
[CV]  max_depth=25, min_samples_leaf=4, n_estimators=300, score=

[CV]  max_depth=30, min_samples_leaf=3, n_estimators=200, score=0.854, total=   0.5s
[CV] max_depth=30, min_samples_leaf=3, n_estimators=200 ..............
[CV]  max_depth=30, min_samples_leaf=3, n_estimators=200, score=0.859, total=   0.5s
[CV] max_depth=30, min_samples_leaf=3, n_estimators=200 ..............
[CV]  max_depth=30, min_samples_leaf=3, n_estimators=200, score=0.854, total=   0.5s
[CV] max_depth=30, min_samples_leaf=3, n_estimators=200 ..............
[CV]  max_depth=30, min_samples_leaf=3, n_estimators=200, score=0.849, total=   0.5s
[CV] max_depth=30, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=30, min_samples_leaf=3, n_estimators=250, score=0.867, total=   0.6s
[CV] max_depth=30, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=30, min_samples_leaf=3, n_estimators=250, score=0.851, total=   0.6s
[CV] max_depth=30, min_samples_leaf=3, n_estimators=250 ..............
[CV]  max_depth=30, min_samples_leaf=3, n_estimators=250, score=

[CV]  max_depth=30, min_samples_leaf=5, n_estimators=300, score=0.851, total=   0.7s


[Parallel(n_jobs=1)]: Done 480 out of 480 | elapsed:  4.0min finished


GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'max_depth': [5, 10, 15, 20, 25, 30],
                         'min_samples_leaf': [2, 3, 4, 5],
                         'n_estimators': [150, 200, 250, 300]},
             verbose=3)

In [284]:
grid.best_params_

{'max_depth': 15, 'min_samples_leaf': 2, 'n_estimators': 250}

In [285]:
rf1 = RandomForestClassifier(n_estimators=250,criterion='gini',max_depth=15,min_samples_split=2,min_samples_leaf=2,
    min_weight_fraction_leaf=0.0,max_features='auto',max_leaf_nodes=None,min_impurity_decrease=0.0,min_impurity_split=None,
    bootstrap=True,
    oob_score=False,
    n_jobs=None,
    random_state=None,
    verbose=0,
    warm_start=False,
    class_weight=None,
    ccp_alpha=0.0,
    max_samples=None,
)

In [286]:
rf1.fit(X_train,y_train)

RandomForestClassifier(max_depth=15, min_samples_leaf=2, n_estimators=250)

In [287]:
predict_rf1 = rf1.predict(X_test)

    Accuracy test for Random Forest after applying Grid Search

In [288]:
print(accuracy_score(y_test,predict_rf1))

0.8308977035490606


In [289]:
print(confusion_matrix(y_test,predict_rf1))

[[678  65]
 [ 97 118]]


In [290]:
print(classification_report(y_test,predict_rf1))

              precision    recall  f1-score   support

       <=50K       0.87      0.91      0.89       743
        >50K       0.64      0.55      0.59       215

    accuracy                           0.83       958
   macro avg       0.76      0.73      0.74       958
weighted avg       0.82      0.83      0.83       958



KNN

In [302]:
from sklearn.neighbors import KNeighborsClassifier

In [328]:
knn = KNeighborsClassifier(n_neighbors=20)

In [329]:
knn.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=20)

In [330]:
predict_knn = knn.predict(X_test)

    Accuracy test for KNN

In [331]:
print(accuracy_score(y_test,predict_knn))

0.8225469728601252


In [332]:
print(confusion_matrix(y_test,predict_knn))

[[671  72]
 [ 98 117]]


In [333]:
print(classification_report(y_test,predict_knn))

              precision    recall  f1-score   support

       <=50K       0.87      0.90      0.89       743
        >50K       0.62      0.54      0.58       215

    accuracy                           0.82       958
   macro avg       0.75      0.72      0.73       958
weighted avg       0.82      0.82      0.82       958



Conclusion

Accuracy of Logistic model is -    0.8298538622129437

Accuracy of Decision Tree is -       0.7943632567849687

Accuracy of Random Forest is - 0.8308977035490606 (same before and after tuning)

Accuracy of KNN is - 0.8225469728601252